internal/decision/engine.go

```diff
@@ -673,12 +673,18 @@ func (e *Engine) ReceiveFrom(from peer.ID, blks []blocks.Block) {
 	// Check each peer to see if it wants one of the blocks we received
 	var work bool
 	missingWants := make(map[peer.ID][]cid.Cid)
-	e.lock.RLock()
 	for _, b := range blks {
 		k := b.Cid()
 
-		for _, p := range e.peerLedger.Peers(k) {
+		e.lock.RLock()
+		peers := e.peerLedger.Peers(k)
+		e.lock.RUnlock()
```

Comment:
Are you assuming that the amount of time to grab the ledger maps for all the peers is long enough that it's worth locking for each peer rather than grabbing all the ledgers at once and chucking them into an array?